### PR TITLE
ConvNetV1

### PR DESCRIPTION
**Pull request by barakugav**

### CODE REVIEW

File: train/data_parser.py [hunk 1]
```diff
@@ -42,18 +42,23 @@ def _unpack_planes_gen(self, nparr_packed_gen):
             planes = [np.unpackbits(plane, count=plane_size)
                       for plane in planes]
             planes = np.array(planes, dtype=np.float32)
+            planes = np.reshape(planes, (-1, self.game.BOARD_SIZE, self.game.BOARD_SIZE))
+            if self.cpu:
+                planes = np.transpose(planes, (1, 2, 0))
```
Comment (poja) [inline train/data_parser.py:48]: change to `NUM_PLANES` or something

File: train/tictactoe.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 
 class NetType:
```
Comment (poja) [inline train/tictactoe.py:18]: Two classes with the same name, should be avoided

File: train/tictactoe.py [hunk 3]
```diff
@@ -40,41 +41,65 @@ def load_data_entry(self, path):
...
+            residual_block_num=cfg["residual_block_num"],
+            moves_num=self.MOVE_NUM)
+        model = Model(inputs=inputs, outputs=outputs)
+
+        opt = optimizers.Adam(learning_rate=_LEARNING_RATE)
+        model.compile(
+            optimizer=opt,
+            loss={'out_value': tf.keras.losses.MeanSquaredError(),
+                  'out_probs': net_utils.loss_cross_entropy},
+            metrics={'out_value': tf.keras.metrics.RootMeanSquaredError(),
```
Comment (poja) [inline train/tictactoe.py:86]: This is taken from the Alpha Zero paper? 
I will have to read a bit of documentation to confirm you on this

**Review by poja — CHANGES_REQUESTED** [review]

File: train/train_process.py [hunk 4]
```diff
@@ -115,6 +115,26 @@ def _train(self, model_path, training_games_dir):
...
+            "--model2-path", model2_path,
+            "--games-num", str(self.cfg["training"]["compare"]["games_num"]),
+            "--result-file", compare_res_file,
+            "--sim-count", str(self.cfg["mcts"]["sim_count"]),
+            "--explore-factor", str(self.cfg["mcts"]["explore_factor"]),
+            "--threads", str(self.cfg["training"]["compare"]["threads"])],
+            stderr=sys.stderr, stdout=sys.stdout, check=True)
+
+        with open(compare_res_file, "r") as res_file:
+            res = json.load(res_file)
```
Comment (poja) [inline train/train_process.py:135]: This file-based communication could be problematic if we have more than one train_process happening at once - but this shouldn't happen, right?

File: train/train_process.py [hunk 5]
```diff
@@ -75,13 +73,14 @@ def run_training_loop(self):
...
-                self.cfg["games_dir"], f"{run_id}_{iter_num}_{model_id}")
+                self.cfg["games_dir"], "{0}_{1:05d}_{2}".format(run_id, iter_num, model_id))
 
             self._self_play(model_path, training_games_dir)
             new_model = self._train(model_path, training_games_dir)
 
             new_model_path = self._save_model(new_model)
-            compare_models(model_path, new_model_path)
+            _w1, _w2, _d = self._compare_models(model_path, new_model_path)
+            # TODO do something with comparison result
```
Comment (poja) [inline train/train_process.py:83]: Let's add something simple for now - 
if _w2 <= _w1, log a warning and don't switch models (continue to the next playing round). 
This will also show us if we have reached rounds with all-ties

File: src/game/net.rs [hunk 6]
```diff
@@ -75,4 +76,11 @@ impl TwoHeadedNetBase {
 
         return (val, probs);
     }
+
+    pub fn softmax_normalizatione(x: Vec<f32>) -> Vec<f32> {
+        let max_p = x.iter().cloned().fold(f32::MIN, f32::max);
```
Comment (poja) [inline src/game/net.rs]: i don't understand where this softmax comes from, i sent you a whatsapp

File: src/hex/net/serializer.rs [hunk 7]
```diff
@@ -34,6 +34,8 @@ impl DataSerializer<HexGame> for HexSerializer {
         planes.push(((pieces_red >> 64) & 0xffffffffffffffff) as u64);
         planes.push(((pieces_blue >> 00) & 0xffffffffffffffff) as u64);
         planes.push(((pieces_blue >> 64) & 0xffffffffffffffff) as u64);
+        /* TODO !!! possible little/big indian bug */
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/hex/net/serializer.rs]: rust doesn't deserialize. we should test this once and not tough this again

File: src/hex/net/two_headed_net.rs [hunk 8]
```diff
@@ -86,11 +86,16 @@ impl TwoHeadedNet {
             .run(&mut args)
             .expect("Error occurred during calculations");
 
-        let val: f32 = args.fetch(output_scalar).unwrap()[0];
+        let mut val: f32 = args.fetch(output_scalar).unwrap()[0];
+        if val.is_nan() {
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/hex/net/two_headed_net.rs]: when using the new conv net, sometimes NaN appears in the output, not sure why.
i don't know if 0 is a good output, what do you think?
for the val maybe it is

File: src/hex/net/two_headed_net.rs [hunk 9]
```diff
@@ -86,11 +86,16 @@ impl TwoHeadedNet {
...
+            val = 0.0;
+        }
 
-        let probs: Tensor<f32> = args.fetch(output_probs).unwrap();
+        let mut probs: Tensor<f32> = args.fetch(output_probs).unwrap();
         for idx in 0..probs.len() {
-            assert!(probs[idx] >= 0.0);
+            if probs[idx].is_nan() {
+                probs[idx] = 0.0;
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/hex/net/two_headed_net.rs]: probebly -1 is better

File: train/tictactoe.py (hunk 3 above)
Comment (barakugav) [reply train/tictactoe.py:86]: image

File: src/hex/net/two_headed_net.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (barakugav) [reply src/hex/net/two_headed_net.rs]: or -infty

File: train/train_process.py (hunk 4 above)
Comment (barakugav) [reply train/train_process.py:135]: for now, but we should use NamedTemporaryFile or something, i never manage to get it work easily

File: src/bin/tictactoe_player_comparator.rs [hunk 10]
```diff
@@ -0,0 +1,66 @@
+use clap::Parser;
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/bin/tictactoe_player_comparator.rs:1]: done in a different commit

File: src/game/net.rs (hunk 6 above)
Comment (barakugav) [reply src/game/net.rs]: replayed to you on whatsapp :)

File: train/train_process.py (hunk 5 above)
Comment (barakugav) [reply train/train_process.py]: added a switching_threshold and warning_threshold

File: src/hex/net/common.rs [hunk 11]
```diff
@@ -31,3 +33,16 @@ pub fn flip_score_if_needed(
...
     }
 }
+
+pub fn position_to_planes(pos: &HexPosition) -> Vec<Bitboard> {
+    let mut planes = Vec::new();
+    /* red pieces plane */
+    planes.push(pos.pieces_red());
+    /* blue pieces plane */
+    planes.push(pos.pieces_blue());
+    /* a plane with all ones to help NN find board edges */
```
Comment (poja) [inline src/hex/net/common.rs:44]: I thought there was supposed to be a plane with ones only in the edges?

File: src/game/net.rs [hunk 12]
```diff
@@ -83,4 +84,33 @@ impl TwoHeadedNetBase {
         let p_sum: f32 = x.iter().sum();
         return x.iter().map(|p| p / p_sum).collect_vec();
     }
+
+    pub fn planes_to_tensor<B: Bitboard>(planes: Vec<B>, board_size: usize) -> Tensor<f32> {
```
Comment (poja) [inline src/game/net.rs]: Rust question - why write like this and not `Vec<Bitboard>`?

File: src/hex/net/common.rs (hunk 11 above)
Comment (barakugav) [reply src/hex/net/common.rs:44]: There is a plane with all ones. Added in another commit in the PR

File: src/hex/net/two_headed_net.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (poja) [reply src/hex/net/two_headed_net.rs]: add a warning every time  these `if`s are activated, and choose whatever value you think
then we can merge this branch and debug the `NaN`s later

File: src/chess/chess_game.rs [hunk 13]
```diff
@@ -206,3 +314,211 @@ impl IGame for ChessGame {
...
+        } else {
+            None
+        };
+        return ChessMove::new(chess::ChessMove::new(source, dest, promotion));
+    })
+    .collect_vec()
+});
+
+static MOVE_TO_NN_INDEX: Lazy<Vec<usize>> = Lazy::new(|| {
+    let mut res = vec![0xffff; 64 * 64 + 22 * 4];
```
Comment (poja) [inline src/chess/chess_game.rs:519]: did not you want something like `usize::MAX` here? instead of 0xffff

File: src/chess/chess_game.rs [hunk 14]
```diff
@@ -24,18 +26,88 @@ impl ChessMove {
...
             Ok(m) => Ok(Self::new(m)),
             Err(e) => Err(err_to_str(e)),
         };
     }
+
+    pub fn get_raw(&self) -> &chess::ChessMove {
+        &self.m
+    }
+
+    pub fn to_idx(&self) -> usize {
```
Comment (poja) [inline src/chess/chess_game.rs:41]: this is sort of a quick hash right? what do you think about calling it `quick_hash`? there are too many things called index - this index, the nn index, and `to_idx` in the ttt/hex meaning of "place on the board"

Comment (poja) [reply src/chess/chess_game.rs:41]: or maybe `serial_number`

File: src/chess/chess_game.rs [hunk 15]
```diff
@@ -24,18 +26,88 @@ impl ChessMove {
...
+        let promotion = self.m.get_promotion();
+        if promotion.is_none() {
+            return self.m.get_source().to_index() * 64 + self.m.get_dest().to_index();
+        } else {
+            let promotions_base = 64 * 64;
+            let sf = self.m.get_source().get_file().to_index();
+            let df = self.m.get_dest().get_file().to_index();
+            let move_chunk = (sf * 2 + df) * 4;
```
Comment (poja) [inline src/chess/chess_game.rs:49]: beseder

File: src/chess/chess_game.rs (hunk 13 above)
Comment (barakugav) [reply src/chess/chess_game.rs:519]: 64*64+22*4=4184= 0x1058, so we are very far from 0xffff, i forgot to change MOVE_TO_NN_INDEX to be Vec<u16>... currently it's 15Kb, which is not ideal

File: src/chess/chess_game.rs [hunk 16]
```diff
@@ -206,3 +314,211 @@ impl IGame for ChessGame {
...
+        };
+        return ChessMove::new(chess::ChessMove::new(source, dest, promotion));
+    })
+    .collect_vec()
+});
+
+static MOVE_TO_NN_INDEX: Lazy<Vec<usize>> = Lazy::new(|| {
+    let mut res = vec![0xffff; 64 * 64 + 22 * 4];
+    for (idx, m) in NN_INDEX_TO_MOVE.iter().enumerate() {
+        res[m.to_idx()] = idx;
```
Comment (poja) [inline src/chess/chess_game.rs:521]: assert empty

File: src/chess/chess_game.rs [hunk 17]
```diff
@@ -53,41 +125,51 @@ impl ChessPosition {
         }
     }
 
-    // pub fn flip_of(pos: &ChessPosition) -> Self {}
+    pub fn flip_of(pos: &ChessPosition) -> Self {
```
Comment (poja) [inline src/chess/chess_game.rs:128]: add a simple flip_twice test?

File: src/chess/chess_game.rs (hunk 15 above)
Comment (barakugav) [reply src/chess/chess_game.rs:49]: this is just the inverse of:
"a7a8q","a7a8r","a7a8b","a7a8k","a7b8q","a7b8r","a7b8b","a7b8k",
"b7a8q","b7a8r","b7a8b","b7a8k","b7b8q","b7b8r","b7b8b","b7b8k",
"b7c8q","b7c8r","b7c8b","b7c8k","c7b8q","c7b8r","c7b8b","c7b8k",
"c7c8q","c7c8r","c7c8b","c7c8k","c7d8q","c7d8r","c7d8b","c7d8k",
"d7c8q","d7c8r","d7c8b","d7c8k","d7d8q","d7d8r","d7d8b","d7d8k",
"d7e8q","d7e8r","d7e8b","d7e8k","e7d8q","e7d8r","e7d8b","e7d8k",
"e7e8q","e7e8r","e7e8b","e7e8k","e7f8q","e7f8r","e7f8b","e7f8k",
"f7e8q","f7e8r","f7e8b","f7e8k","f7f8q","f7f8r","f7f8b","f7f8k",
"f7g8q","f7g8r","f7g8b","f7g8k","g7f8q","g7f8r","g7f8b","g7f8k",
"g7g8q","g7g8r","g7g8b","g7g8k","g7h8q","g7h8r","g7h8b","g7h8k",
"h7g8q","h7g8r","h7g8b","h7g8k","h7h8q","h7h8r","h7h8b","h7h8k"

File: src/chess/chess_game.rs (hunk 14 above)
Comment (barakugav) [reply src/chess/chess_game.rs:41]: Hash is not what we need, we need a fixed indexing, which doesn't span over a huge range, as MOVE_TO_NN_INDEX size is defined by the max move index.
We can rename some things, maybe call this to_int

File: src/chess/chess_game.rs (hunk 17 above)
Comment (barakugav) [reply src/chess/chess_game.rs:128]: will add

File: src/chess/chess_game.rs (hunk 14 above)
Comment (poja) [reply src/chess/chess_game.rs:41]: now i see how it's similar to the ttt/hex definition of `idx`

File: train/chess.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+            cpu=cfg["cpu"])
+        model = Model(inputs=inputs, outputs=outputs)
+
+        opt = optimizers.Adam(learning_rate=_LEARNING_RATE)
+        model.compile(
+            optimizer=opt,
+            loss={'value_head': tf.keras.losses.MeanSquaredError(),
+                  'policy_head': net_utils.loss_cross_entropy},
+            metrics={'value_head': tf.keras.metrics.RootMeanSquaredError(),
+                     'policy_head': net_utils.policy_head_accuracy})
```
Comment (poja) [inline train/chess.py:86]: how do you choose the metrics? they are just for human-reading right? they don't affect the training process

Comment (barakugav) [reply train/chess.py:86]: yes, but they are the best indication we have. I also want to open a ticket of collecting the return value of model.fit, which return some sort of history with the metrics. This will give us a lot of debug information
